File: src/core/thermalized_bond.cpp
/*
  Copyright (C) 2010,2011,2012,2013 The ESPResSo project
  Copyright (C) 2002,2003,2004,2005,2006,2007,2008,2009,2010 
    Max-Planck-Institute for Polymer Research, Theory Group
  
  This file is part of ESPResSo.
  
  ESPResSo is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  ESPResSo is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/
/** \file thermalized_bond.cpp
 *
 *  Implementation of \ref thermalized_bond.hpp
 */

#include "thermalized_bond.hpp"
#include "communication.hpp"
#include "global.hpp"
#include "interaction_data.hpp"

int n_thermalized_bonds = 0;

int thermalized_bond_set_params(int bond_type, double temp_com, double gamma_com, double temp_distance, double gamma_distance, double r_cut)
{
  if (bond_type < 0)
    return ES_ERROR;

  make_bond_type_exist(bond_type);

  bonded_ia_params[bond_type].p.thermalized_bond.temp_com = temp_com;
  bonded_ia_params[bond_type].p.thermalized_bond.gamma_com = gamma_com;
  bonded_ia_params[bond_type].p.thermalized_bond.temp_distance = temp_distance;
  bonded_ia_params[bond_type].p.thermalized_bond.gamma_distance = gamma_distance;
  bonded_ia_params[bond_type].p.thermalized_bond.r_cut = r_cut;
 
  bonded_ia_params[bond_type].p.thermalized_bond.pref1_com = gamma_com;
  bonded_ia_params[bond_type].p.thermalized_bond.pref2_com = sqrt(24.0 * gamma_com / time_step * temp_com);
  bonded_ia_params[bond_type].p.thermalized_bond.pref1_dist = gamma_distance;
  bonded_ia_params[bond_type].p.thermalized_bond.pref2_dist = sqrt(24.0 * gamma_distance / time_step * temp_distance);

  bonded_ia_params[bond_type].type = BONDED_IA_THERMALIZED_DIST;

  bonded_ia_params[bond_type].num = 1;

  n_thermalized_bonds += 1;
  mpi_bcast_ia_params(bond_type, -1); 
  mpi_bcast_parameter(FIELD_THERMALIZEDBONDS);

  return ES_OK;
}

void thermalized_bond_heat_up() {
  double pref_scale = sqrt(3);
  thermalized_bond_update_params(pref_scale);
}

void thermalized_bond_cool_down() {
  double pref_scale = 1.0 / sqrt(3);
  thermalized_bond_update_params(pref_scale);
}

void thermalized_bond_init()
{

  for (int i = 0; i < bonded_ia_params.size(); i++) {
     if (bonded_ia_params[i].type == BONDED_IA_THERMALIZED_DIST) {
        Thermalized_bond_parameters &t = bonded_ia_params[i].p.thermalized_bond;
        t.pref1_com = t.gamma_com;
        t.pref2_com = sqrt(24.0 * t.gamma_com / time_step * t.temp_com);
        t.pref1_dist = t.gamma_distance;
        t.pref2_dist = sqrt(24.0 * t.gamma_distance / time_step * t.temp_distance);
    }
  }
}


void thermalized_bond_update_params(double pref_scale) {
  
  for (int i = 0; i < bonded_ia_params.size(); i++) {
     if (bonded_ia_params[i].type == BONDED_IA_THERMALIZED_DIST) {
        Thermalized_bond_parameters &t = bonded_ia_params[i].p.thermalized_bond;
        t.pref2_com *= pref_scale;
        t.pref2_dist *= pref_scale;
    }
  }
}

File: testsuite/python/thermalized_bond.py
#
# Copyright (C) 2013-2022 The ESPResSo project
#
# This file is part of ESPResSo.
#
# ESPResSo is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ESPResSo is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
import numpy as np
import unittest as ut
import unittest_decorators as utx
import thermostats_common

import espressomd
import espressomd.interactions


@utx.skipIfMissingFeatures(["MASS"])
class ThermalizedBond(ut.TestCase, thermostats_common.ThermostatsCommon):

    """
    Test the two velocity distributions for COM and distance created by the
    thermalized bond independently against the single component Maxwell
    distribution.
    """

    system = espressomd.System(box_l=[10., 10., 10.])
    system.cell_system.set_n_square()
    system.cell_system.skin = 0.3

    def setUp(self):
        np.random.seed(42)
        self.system.time_step = 0.01
        self.system.periodicity = [True, True, True]

    def tearDown(self):
        self.system.part.clear()
        self.system.thermostat.turn_off()

    def test_com_langevin(self):
        """Test for COM thermalization."""

        N = 200
        N_half = int(N / 2)
        self.system.time_step = 0.02
        self.system.periodicity = [False, False, False]

        m1 = 1.0
        m2 = 10.0
        # Place particles

        partcls_m1 = self.system.part.add(
            pos=np.random.random(
                (N_half, 3)), mass=N_half * [m1])
        partcls_m2 = self.system.part.add(
            pos=np.random.random(
                (N_half, 3)), mass=N_half * [m2])

        t_dist = 0
        g_dist = 0
        t_com = 2.0
        g_com = 4.0

        self.system.thermostat.set_thermalized_bond(seed=55)
        thermalized_dist_bond = espressomd.interactions.ThermalizedBond(
            temp_com=t_com, gamma_com=g_com, temp_distance=t_dist,
            gamma_distance=g_dist, r_cut=2.0)
        self.system.bonded_inter.add(thermalized_dist_bond)

        for p1, p2 in zip(partcls_m1, partcls_m2):
            p1.add_bond((thermalized_dist_bond, p2))

        # Warmup
        self.system.integrator.run(50)

        # Sampling
        loops = 150
        v_stored = np.zeros((N_half * loops, 3))
        for i in range(loops):
            self.system.integrator.run(5)
            v_com = 1.0 / \
                (m1 + m2) * \
                (m1 * partcls_m1.v + m2 * partcls_m2.v)
            v_stored[i * N_half:(i + 1) * N_half, :] = v_com

        v_minmax = 5
        bins = 50
        error_tol = 0.017
        self.check_velocity_distribution(
            v_stored, v_minmax, bins, error_tol, t_com)

    def test_dist_langevin(self):
        """Test for dist thermalization."""

        N = 100
        N_half = int(N / 2)
        self.system.time_step = 0.02
        self.system.periodicity = [True, True, True]

        m1 = 1.0
        m2 = 10.0
        # Place particles
        partcls_m1 = self.system.part.add(
            pos=np.random.random(
                (N_half, 3)), mass=N_half * [m1])
        partcls_m2 = self.system.part.add(
            pos=np.random.random(
                (N_half, 3)), mass=N_half * [m2])

        t_dist = 2.0
        g_dist = 4.0
        t_com = 0.0
        g_com = 0.0

        self.system.thermostat.set_thermalized_bond(seed=51)
        thermalized_dist_bond = espressomd.interactions.ThermalizedBond(
            temp_com=t_com, gamma_com=g_com, temp_distance=t_dist,
            gamma_distance=g_dist, r_cut=9)
        self.system.bonded_inter.add(thermalized_dist_bond)

        for p1, p2 in zip(partcls_m1, partcls_m2):
            p1.add_bond((thermalized_dist_bond, p2))

        # Warmup
        self.system.integrator.run(50)

        # Sampling
        loops = 150
        v_stored = np.zeros((N_half * loops, 3))
        for i in range(loops):
            self.system.integrator.run(5)
            v_dist = partcls_m2.v - partcls_m1.v
            v_stored[i * N_half:(i + 1) * N_half, :] = v_dist

        v_minmax = 5
        bins = 50
        error_tol = 0.015
        self.check_velocity_distribution(
            v_stored, v_minmax, bins, error_tol, t_dist)

    def test_exceptions(self):
        espressomd.interactions.ThermalizedBond(
            temp_com=1., gamma_com=1., temp_distance=1., gamma_distance=1.,
            r_cut=3.)
        with self.assertRaisesRegex(Exception, "Thermalized bonds require the thermalized_bond thermostat"):
            self.system.integrator.run(0)


if __name__ == "__main__":
    ut.main()
